src/pybind/mgr/devicehealth/module.py

```diff
@@ -273,44 +305,22 @@ def shutdown(self):
         self.event.set()
 
     def open_connection(self, create_if_missing=True):
-        pools = self.rados.list_pools()
-        is_pool = False
-        for pool in pools:
-            if pool == self.pool_name:
-                is_pool = True
-                break
-        if not is_pool:
-            if not create_if_missing:
-                return None
-            self.log.debug('create %s pool' % self.pool_name)
-            # create pool
-            result = CommandResult('')
-            self.send_command(result, 'mon', '', json.dumps({
-                'prefix': 'osd pool create',
-                'format': 'json',
-                'pool': self.pool_name,
-                'pg_num': 1,
-                'pg_num_min': 1,
-            }), '')
-            r, outb, outs = result.wait()
-            assert r == 0
-
-            # set pool application
-            result = CommandResult('')
-            self.send_command(result, 'mon', '', json.dumps({
-                'prefix': 'osd pool application enable',
-                'format': 'json',
-                'pool': self.pool_name,
-                'app': 'mgr_devicehealth',
-            }), '')
-            r, outb, outs = result.wait()
-            assert r == 0
-
+        if not create_if_missing:
+            return None
```

Comment:
This leads to `self.open_connection(create_if_missing=False)` in `_get_device_metrics()`
to always return `None`, and therefore `_get_device_metrics` will always return an empty dictionary.
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/devicehealth/module.py#L426

This is the reason the `assert` failed in `self_test()`:
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/devicehealth/module.py#L214

Reply:
Thanks @yaarith.
But still I see `self_test` is failing on master as well.

Reply:
@sunnyku I tested it on master yesterday and `self_test` worked;

Are you still seeing the errors mentioned here:
https://gist.github.com/sunnyku/044254062c061d386a0bb82f6a2aab90
?

The JSON error is probably due to an old version of `smartctl` ( < 7.0 ).

Reply:
@sunnyku 

Rerun it again on master now:
```
$ ceph mgr self-test module devicehealth
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-09-25T13:25:25.484+0000 7f3ab8a94700 -1 WARNING: all dangerous and experimental features are enabled.
2020-09-25T13:25:25.581+0000 7f3ab8a94700 -1 WARNING: all dangerous and experimental features are enabled.
Self-test OK
None
```

Reply:
➜  build git:(wip-fix-health) ./bin/ceph mgr self-test module devicehealth                         
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-09-25T14:29:44.614+0100 7fd037be3700 -1 WARNING: all dangerous and experimental features are enabled.
2020-09-25T14:29:44.636+0100 7fd037be3700 -1 WARNING: all dangerous and experimental features are enabled.
Self-test OK
None
➜  build git:(wip-fix-health) ./bin/ceph mgr self-test module devicehealth
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-09-25T14:29:49.492+0100 7fa0adb57700 -1 WARNING: all dangerous and experimental features are enabled.
2020-09-25T14:29:49.514+0100 7fa0adb57700 -1 WARNING: all dangerous and experimental features are enabled.
Error EPERM: Test failed: Remote method threw exception: Traceback (most recent call last):
  File "/home/sunny/lab/ceph/src/pybind/mgr/devicehealth/module.py", line 214, in self_test
    assert before != after
AssertionError

➜  build git:(wip-fix-health) ./bin/ceph mgr self-test module devicehealth
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-09-25T14:29:59.572+0100 7f00b56ad700 -1 WARNING: all dangerous and experimental features are enabled.
2020-09-25T14:29:59.594+0100 7f00b56ad700 -1 WARNING: all dangerous and experimental features are enabled.
Error EPERM: Test failed: Remote method threw exception: Traceback (most recent call last):
  File "/home/sunny/lab/ceph/src/pybind/mgr/devicehealth/module.py", line 214, in self_test
    assert before != after
AssertionError

➜  build git:(wip-fix-health) ./bin/ceph mgr self-test module devicehealth
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-09-25T14:30:09.523+0100 7f470faca700 -1 WARNING: all dangerous and experimental features are enabled.
2020-09-25T14:30:09.547+0100 7f470faca700 -1 WARNING: all dangerous and experimental features are enabled.
Self-test OK
None

Reply:
It's not consistent, I have raised https://tracker.ceph.com/issues/47645.

Reply:
Also my box has `smartctl 7.1`

Reply:
@sunnyku 
Which code is it running? I see wip-fix-health

Reply:
Its master only just brach name is different.

master         e066050636 [origin/master] Merge pull request #37136 from p-se/cd-add-per-rbd-graphs
wip-fix-health e066050636 [origin/master] Merge pull request #37136 from p-se/cd-add-per-rbd-graphs

Reply:
> Also my box has `smartctl 7.1`

In this case the error might be related to the one causing the assertion error.

Is your box a VM by any chance?

Reply:
No, just regular work laptop with Fedora-32.

Reply:
I followed the steps to reproduce the bug, but it keeps outputing `Self-test OK`.

Is there a public run where I can take a look at the logs?

Reply:
@yaarith See the description and first comment on https://tracker.ceph.com/issues/46743 for a more thorough description of the bug this is fixing. It should be trivial to reproduce by running `cephadm bootstrap` and then refraining from deploying any OSDs.

That said, I posted a MGR log dump here: https://paste2.org/vMj9cPCD